cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/JarCacheTracker.java

```diff
@@ -0,0 +1,173 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardWatchEventKinds;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Optimization for CDAP-7021.
+ *
+ * We cache appmaster and container jars created by twill the first time a program
+ * of that type is run. However, if 100 programs are started at the same time, all 100 will still create those
+ * jars since the cached jars won't be available at run start. So as an optimization, if a program is
+ * launching and its cached jars are not available, and its not the first program of its type to be launched,
+ * and its launching within 90 seconds of the first run, wait for the cache jars to become available.
+ * Without this, we might actually miss schedule runs due to jar building taking too much time and resources.
+ */
+public class JarCacheTracker {
+  public static final JarCacheTracker INSTANCE = new JarCacheTracker();
+  private static final Logger LOG = LoggerFactory.getLogger(JarCacheTracker.class);
+  private final Map<ProgramType, CacheInfo> programCaches;
+
+  private JarCacheTracker() {
+    programCaches = new ConcurrentHashMap<>();
+    for (ProgramType programType : ProgramType.values()) {
+      programCaches.put(programType, new CacheInfo());
+    }
+  }
+
+  public void registerLaunch(File cacheDir, ProgramType programType) {
+    CacheInfo cacheInfo = programCaches.get(programType);
+
+    long launchTime = System.currentTimeMillis();
+    if (cacheInfo.firstLaunchTime.compareAndSet(-1L, launchTime)) {
+      // if this is the first program of this type to launch, no need to wait
+      LOG.debug("First program of type {} launched at time {}", programType, launchTime);
+      return;
+    }
+
+    // if we're more than 90 seconds past the first launch time, no need to waits
+    long threshold = cacheInfo.firstLaunchTime.get() + TimeUnit.MILLISECONDS.convert(90, TimeUnit.SECONDS);
+    if (System.currentTimeMillis() > threshold) {
+      return;
+    }
+
+    // otherwise, wait until at most 90 seconds past the first launch time for the cache files to exist
+    cacheInfo.waitForCacheFiles(cacheDir);
+  }
+
+  /**
+   * Used to hold the first time a program of the type was launched and wait for the cache files to exist.
+   */
+  private class CacheInfo {
+    private final AtomicLong firstLaunchTime = new AtomicLong(-1L);
+
+    // synchronized here instead of at registerLaunch so that we only per program type.
+    private synchronized void waitForCacheFiles(File cacheDir) {
+      Path cachePath = cacheDir.getAbsoluteFile().toPath();
+      Path containerDonePath = cachePath.resolve("container.jar.done");
```

Comment:
I think it would be a good idea to use constants (`Constants.Files.APP_MASTER_JAR`) here.